### DISCUSSION TITLE
Octal integer literals

### DISCUSSION THREAD

**Discussion by colejohnson66** [dotnet/csharplang#7886] [Language Ideas]
Sometimes, when writing P/Invoke code, we encounter enum-like constants that are written in C using octal literals. For example, [from `<bits/poll.h>` in glibc](https://github.com/bminor/glibc/blob/glibc-2.38/bits/poll.h#L37-L42), `POLLERR`, `POLLHUP`, and `POLLNVAL` are written using C's octal literal notation (prefixed by a lone zero). These *can* be translated into C# by writing their equivalent decimal, hexadecimal, or binary value, but this is error-prone.

I propose that C# support octal literals, but with a caveat. In C, it is widely considered a mistake that any zero-prefixed number is octal. Instead, we should require a `0o` or `0O` (`U+0030 DIGIT ZERO` followed by either a `U+004F LATIN CAPITAL LETTER O` or `U+006F LATIN SMALL LETTER O`) prefix, similar to hexadecimal. There is some precedent for this from multiple languages:

* [ECMAScript](https://262.ecma-international.org/14.0/#prod-OctalIntegerLiteral)
* [Haskell](https://www.haskell.org/onlinereport/lexemes.html#sect2.5)
* [NASM](https://www.nasm.us/doc/nasmdoc3.html#section-3.4.1)
* [PHP](https://www.php.net/manual/en/language.types.integer.php)
* [Python](https://docs.python.org/3.1/reference/lexical_analysis.html#integer-literals)
* [Rust](https://doc.rust-lang.org/rust-by-example/primitives/literals.html)
* ...and more

As with other integer literals, underscore-based separators would be allowed.

The language grammar would be altered as follows:
```diff
 // 6.4.5.3 Integer literals
 Integer_Literal
     : Decimal_Integer_Literal
     | Hexadecimal_Integer_Literal
     | Binary_Integer_Literal
+    | Octal_Integer_Literal
     ;

+fragment Octal_Integer_Literal
+    : ('0o' | '0O') Decorated_Octal_Digit* Integer_Type_Suffix?
+    ;

+fragment Decorated_Octal_Digit
+    : '_'* Octal_Digit
+    ;

+fragment Octal_Digit
+    : '0'..'7'
+    ;
```

Alternatively, a different letter can be chosen to avoid visual collision between `U+0030 DIGIT ZERO` and `U+004F LATIN CAPITAL LETTER O`. NASM resolves this potential issue by also allowing `U+0072 LATIN SMALL LETTER Q` as a separator, but this could cause confusion.

**Comment by HaloFour** [comment]
I remember this coming up during the discussion around [binary literals](https://github.com/dotnet/roslyn/issues/215#issuecomment-73151243) but I don't see it mentioned in any of the meeting notes.

**Reply by colejohnson66** [reply]
Binary literals are definitely more useful than octal, so it makes sense for them to have been prioritized (assuming that's what happened).

**Comment by TahirAhmadov** [comment]
Why not 0C/0c?

**Reply by DavidArno** [reply]
That sounds a bit like bikeshedding. Surely whether it's `0c`, `0o` or `Ook` isn't as important as the desire to express numbers using octal?

**Reply by TahirAhmadov** [reply]
Not really, 1, I support the proposal and "liked" it right away, 2, `0O/0o` looks weird and when adding a feature everything has to be done as close to perfectly as possible.

**Comment by huoyaoyuan** [comment]
Group separated binary literal can somehow solve it. For example, `0b_010_000`.
